src/rnnlm/rnnlm-component.cc

```diff
@@ -0,0 +1,788 @@
+//  bias_params_.SetRandn();
+//  bias_params_.Scale(bias_stddev);
+}
+
+void AffineSampleLogSoftmaxComponent::Init(std::string matrix_filename) {
+  Matrix<BaseFloat> mat;
+  ReadKaldiObject(matrix_filename, &mat); // will abort on failure.
+  KALDI_ASSERT(mat.NumCols() >= 2);
+  int32 input_dim = mat.NumCols() - 1, output_dim = mat.NumRows();
+  linear_params_.Resize(output_dim, input_dim);
+  bias_params_.Resize(output_dim);
+  linear_params_.CopyFromMat(mat.Range(0, output_dim, 0, input_dim));
+  bias_params_.CopyColFromMat(mat, input_dim);
+}
+
+void AffineSampleLogSoftmaxComponent::InitFromConfig(ConfigLine *cfl) {
+  bool ok = true;
+  std::string matrix_filename;
+  int32 input_dim = -1, output_dim = -1;
+  InitLearningRatesFromConfig(cfl);
+  if (cfl->GetValue("matrix", &matrix_filename)) {
+    Init(matrix_filename);
+    if (cfl->GetValue("input-dim", &input_dim))
+      KALDI_ASSERT(input_dim == InputDim() &&
+                   "input-dim mismatch vs. matrix.");
+    if (cfl->GetValue("output-dim", &output_dim))
+      KALDI_ASSERT(output_dim == OutputDim() &&
+                   "output-dim mismatch vs. matrix.");
+  } else {
+    ok = ok && cfl->GetValue("input-dim", &input_dim);
+    ok = ok && cfl->GetValue("output-dim", &output_dim);
+//    BaseFloat param_stddev = 1.0 / std::sqrt(input_dim),
+//        bias_stddev = 1.0;
+    BaseFloat param_stddev = 0.0, /// log(1.0 / output_dim),
+        bias_stddev = log(1.0 / output_dim);
+    cfl->GetValue("param-stddev", &param_stddev);
+    cfl->GetValue("bias-stddev", &bias_stddev);
+    Init(input_dim, output_dim, param_stddev, bias_stddev);
+  }
+  if (cfl->HasUnusedValues())
+    KALDI_ERR << "Could not process these elements in initializer: "
+              << cfl->UnusedValues();
+  if (!ok)
+    KALDI_ERR << "Bad initializer " << cfl->WholeLine();
+}
+
+void AffineSampleLogSoftmaxComponent::Propagate(const MatrixBase<BaseFloat> &in,
+                                                const vector<vector<int> > &indexes,
+                                                vector<vector<BaseFloat> > *out) const {
+//  KALDI_LOG << "sum is " << bias_params_.Sum();
+//  KALDI_ASSERT(bias_params_.Sum() == bias_params_.Sum());
+//  KALDI_ASSERT(in.NumRows() == indexes.size());
+  out->resize(indexes.size());
+
+  for (int i = 0; i < indexes.size(); i++) {
+    (*out)[i].resize(indexes[i].size());
+
+    for (int j = 0; j < indexes[i].size(); j++) {
+      int w = indexes[i][j];
+      BaseFloat res = VecVec(in.Row(i), linear_params_.Row(w));
+      (*out)[i][j] = res + bias_params_(w);
+      KALDI_ASSERT((*out)[i][j] == (*out)[i][j]);
+    }
+  }
+}
+
+void AffineSampleLogSoftmaxComponent::Backprop(
+                               const vector<vector<int> > &indexes,
+                               const MatrixBase<BaseFloat> &in_value,
+                               const MatrixBase<BaseFloat> &, // out_value
+                               const vector<vector<BaseFloat> > &output_deriv,
```

Comment:
vector<vector<BaseFloat> > is the wrong data-structure to use here...
all of this stuff should be done on GPU, you can use functions with names *CopyRows*() to selectively copy rows from a matrix.